source_code/main_mcu/src/fido2/fido2.c

```diff
@@ -0,0 +1,493 @@
+
+/*! \fn     fido2_update_credential_count(uint32_t count)
+*   \brief  Update the credential count. This is to protect against replay attack.
+*           Server will check that count is not going backwards or re-used.
+*   \param  Current count to be updated
+*   \return new count
+*/
+static uint32_t fido2_update_credential_count(uint32_t count)
+{
+    //TBD
+    count++;
+    return count;
+}
+
+/*! \fn     fido2_prompt_user(char const *prompt, bool wait_for_ack, uint32_t timeout_sec)
+*   \brief  Prompt the user and optionally wait for ack from user.
+*   \param  prompt to display
+*   \param  boolean whether to wait for ACK
+*   \param  the amount of time to display message/wait for ack before timeout
+*   \return 0 if accepted or other value if denied/failed
+*/
+static uint32_t fido2_prompt_user(char const *prompt, bool wait_for_ack, uint32_t timeout_sec)
+{
+    uint32_t result = SUCCESS;
+    //TBD
+    //Display request to create cred with name
+    //Wait for ack
+    return result;
+}
+
+/*! \fn     fido2_make_auth_tag(credential_ID_t *cred_ID)
+*   \brief  Make a tag uniquely identifying credential
+*   \param  credential id
+*   \return void
+*/
+static void fido2_make_auth_tag(credential_ID_t *cred_ID)
+{
+    rng_fill_array(cred_ID->tag, TAG_LEN);
+}
+
+/*! \fn     fido2_calc_attestation_signature(uint8_t const *data, int datalen, uint8_t const * client_data_hash, uint8_t * sigbuf)
+*   \brief  Calculate the attestation signature
+*   \param  data to calculate signature over
+*   \param  length of data
+*   \param  hash of client data
+*   \param  out buffer with signature
+*   \return void
+*/
+static void fido2_calc_attestation_signature(uint8_t const *data, int datalen, uint8_t const * client_data_hash, uint8_t * sigbuf)
+{
+    #define HASH_LEN 32
+    uint8_t hashbuf[HASH_LEN];
+
+    crypto_sha256_init();
+    crypto_sha256_update(data, datalen);
+    crypto_sha256_update(client_data_hash, CLIENT_DATA_HASH_LEN);
+    crypto_sha256_final(hashbuf);
+
+    crypto_ecc256_sign(hashbuf, HASH_LEN, sigbuf);
+}
+
+/*! \fn     fido2_cbor_encode_public_key(uint8_t *buf, uint32_t bufLen, ecc256_pub_key const *pub_key)
+*   \brief  CBOR encoder for the public key. All cbor encoding and decoding belongs to aux_mcu.
+*           Unfortunately, to be able to create attestation signature we have to encode the public key
+*           in cbor format so we have to do some cbor encoding on main_mcu also.
+*           Essentiallyt the same as Solo function ctap_add_coseKey() but
+*           removed error checking and pass in buffer instead of the cbor encoder object
+*           in addition to hardcoding algtype.
+*   \param  buffer for encoded output
+*   \param  buffer length
+*   \param  public key
+*   \return encoded public key buffer size
+*/
+static uint32_t fido2_cbor_encode_public_key(uint8_t *buf, uint32_t bufLen, ecc256_pub_key const *pub_key)
+{
+    CborEncoder cose_key;
+    CborEncoder map;
```

Comment:
how much RAM does both these vars take btw?

Reply:
Stack you mean? I am not sure.

Reply:
yup :)